primitives/runtime/src/traits.rs

```diff
@@ -1210,11 +1215,42 @@ pub trait Header: Clone + Send + Sync + Codec + Eq + MaybeSerialize + Debug + 's
 	}
 }
 
+/// Something that provides the Header Type.
+// This is needed to fix the "cyclical" issue in loading Header/BlockNumber as part of a
+// `pallet::call`. Essentially, `construct_runtime` aggregates all calls to create a `RuntimeCall`
+// that is then used to define `UncheckedExtrinsic`.
+// ```ignore
+// pub type UncheckedExtrinsic =
+// 	generic::UncheckedExtrinsic<Address, RuntimeCall, Signature, SignedExtra>;
+// ```
+// This `UncheckedExtrinsic` is supplied to the `Block`.
+// ```ignore
+// pub type Block = generic::Block<Header, UncheckedExtrinsic>;
+// ```
+// So, if we do not create a trait outside of `Block` that doesn't have `Extrinsic`, we go into a
+// recursive loop leading to a build error.
+//
+// Note that this is a workaround and should be removed once we have a better solution.
+pub trait HeaderProvider {
+	/// Header type.
+	type HeaderT: Header;
```

Comment:
A naming nit, usually in substrate we are fine with something like `type Header: Header` and later on, even if we are to rename one of them to something else to avoid collision, we do `use traits::Header as HeaderT`, so the other way around.

Reply:
Yes, this was done to avoid collision with `Header` in `Block`. Else, we will need to manually cast `<SomeBlock as Block>::Header` to access it in all places to avoid ambiguity.